File: samples/microsoft/csharp/enterprise-agent-tutorial/2-prototype-to-production/ProductionDeployment.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text.Json;
using System.Threading.Tasks;

namespace Microsoft.Azure.Samples.PrototypeToProduction
{
    /// <summary>
    /// Production Deployment module for endpoint configuration and external consumption.
    /// This module provides:
    /// - Container Apps endpoint configuration
    /// - Azure Marketplace integration
    /// - Logic Apps connector setup
    /// - Managed identity and authentication
    /// - API Management integration
    /// </summary>
    public class ProductionDeployment
    {
        /// <summary>
        /// Creates production endpoint configuration for deploying the agent.
        /// </summary>
        /// <param name="agentName">Name of the agent to deploy</param>
        /// <returns>Endpoint configuration dictionary</returns>
        public async Task<Dictionary<string, object>> CreateProductionEndpointAsync(string agentName)
        {
            Console.WriteLine("\n🚀 PRODUCTION ENDPOINT DEPLOYMENT");
            Console.WriteLine(new string('=', 70));

            // <endpoint_configuration>
            // NOTE: This code is a non-runnable snippet of the larger sample code from which it is taken.
            var endpointConfig = new Dictionary<string, object>
            {
                ["timestamp"] = DateTime.UtcNow.ToString("O"),
                ["agent_name"] = agentName,
                ["endpoint"] = new Dictionary<string, object>
                {
                    ["type"] = "container_app",
                    ["url"] = $"https://{agentName.ToLower().Replace(" ", "-")}.azurecontainerapps.io",
                    ["api_version"] = "2025-05-15-preview",
                    ["authentication"] = "managed_identity"
                },
                ["marketplace"] = new Dictionary<string, object>
                {
                    ["enabled"] = true,
                    ["listing_name"] = "Modern Workplace Assistant",
                    ["category"] = "AI + Machine Learning",
                    ["pricing_model"] = "consumption_based",
                    ["description"] = "Enterprise AI assistant combining internal knowledge with external guidance"
                },
                ["logic_apps"] = new Dictionary<string, object>
                {
                    ["enabled"] = true,
                    ["connector_name"] = "workplace-assistant-connector",
                    ["actions"] = new[] {
                        "ask_question",
                        "search_policy",
                        "get_technical_guidance"
                    },
                    ["triggers"] = new[] {
                        "new_question_received",
                        "policy_update_needed"
                    }
                },
                ["identity"] = new Dictionary<string, object>
                {
                    ["type"] = "managed_identity",
                    ["principal_id"] = "auto-generated",
                    ["tenant_id"] = Environment.GetEnvironmentVariable("AZURE_TENANT_ID") ?? "your-tenant-id",
                    ["permissions"] = new[] {
                        "Agent.Read",
                        "Agent.Execute",
                        "Conversation.Create",
                        "Response.Create"
                    }
                },
            // </endpoint_configuration>
                ["api_management"] = new Dictionary<string, object>
                {
                    ["enabled"] = true,
                    ["rate_limiting"] = new Dictionary<string, object>
                    {
                        ["requests_per_minute"] = 100,
                        ["requests_per_hour"] = 5000
                    },
                    ["cors"] = new Dictionary<string, object>
                    {
                        ["enabled"] = true,
                        ["allowed_origins"] = new[] { "https://*.microsoft.com", "https://*.azure.com" }
                    },
                    ["api_key_required"] = true
                }
            };

            var endpoint = (Dictionary<string, object>)endpointConfig["endpoint"];
            var marketplace = (Dictionary<string, object>)endpointConfig["marketplace"];
            var logicApps = (Dictionary<string, object>)endpointConfig["logic_apps"];
            var identity = (Dictionary<string, object>)endpointConfig["identity"];

            Console.WriteLine("🌐 Endpoint Configuration:");
            Console.WriteLine($"   Type: {endpoint["type"]}");
            Console.WriteLine($"   URL: {endpoint["url"]}");
            Console.WriteLine($"   Authentication: {endpoint["authentication"]}");

            Console.WriteLine($"\n📦 Marketplace Integration:");
            Console.WriteLine($"   Status: {((bool)marketplace["enabled"] ? "✅ ENABLED" : "❌ DISABLED")}");
            Console.WriteLine($"   Listing: {marketplace["listing_name"]}");
            Console.WriteLine($"   Pricing: {marketplace["pricing_model"]}");

            Console.WriteLine($"\n🔗 Logic Apps Integration:");
            Console.WriteLine($"   Status: {((bool)logicApps["enabled"] ? "✅ ENABLED" : "❌ DISABLED")}");
            Console.WriteLine($"   Connector: {logicApps["connector_name"]}");
            Console.WriteLine($"   Actions: {((string[])logicApps["actions"]).Length}");
            Console.WriteLine($"   Triggers: {((string[])logicApps["triggers"]).Length}");

            Console.WriteLine($"\n🔐 Identity & Access:");
            Console.WriteLine($"   Type: {identity["type"]}");
            Console.WriteLine($"   Permissions: {((string[])identity["permissions"]).Length}");

            var logicAppDefinition = new Dictionary<string, object>
            {
                ["definition"] = new Dictionary<string, object>
                {
                    ["$schema"] = "https://schema.management.azure.com/providers/Microsoft.Logic/schemas/2016-06-01/workflowdefinition.json#",
                    ["actions"] = new Dictionary<string, object>
                    {
                        ["Query_Workplace_Assistant"] = new Dictionary<string, object>
                        {
                            ["type"] = "Http",
                            ["inputs"] = new Dictionary<string, object>
                            {
                                ["method"] = "POST",
                                ["uri"] = $"{endpoint["url"]}/api/query",
                                ["headers"] = new Dictionary<string, string>
                                {
                                    ["Content-Type"] = "application/json"
                                },
                                ["body"] = new Dictionary<string, string>
                                {
                                    ["question"] = "@triggerBody()['question']",
                                    ["conversation_id"] = "@triggerBody()['conversation_id']"
                                },
                                ["authentication"] = new Dictionary<string, string>
                                {
                                    ["type"] = "ManagedServiceIdentity"
                                }
                            }
                        }
                    },
                    ["triggers"] = new Dictionary<string, object>
                    {
                        ["manual"] = new Dictionary<string, object>
                        {
                            ["type"] = "Request",
                            ["kind"] = "Http",
                            ["inputs"] = new Dictionary<string, object>
                            {
                                ["schema"] = new Dictionary<string, object>
                                {
                                    ["type"] = "object",
                                    ["properties"] = new Dictionary<string, object>
                                    {
                                        ["question"] = new Dictionary<string, string> { ["type"] = "string" },
                                        ["conversation_id"] = new Dictionary<string, string> { ["type"] = "string" }
                                    }
                                }
                            }
                        }
                    }
                }
            };

            await File.WriteAllTextAsync("endpoint_config.json",
                JsonSerializer.Serialize(endpointConfig, new JsonSerializerOptions { WriteIndented = true }));
            await File.WriteAllTextAsync("logic_app_definition.json",
                JsonSerializer.Serialize(logicAppDefinition, new JsonSerializerOptions { WriteIndented = true }));

            Console.WriteLine($"\n✅ Production endpoint configured successfully");
            Console.WriteLine($"   📄 Configuration saved to: endpoint_config.json");
            Console.WriteLine($"   📄 Logic App definition: logic_app_definition.json");
            Console.WriteLine($"\n📚 Next Steps:");
            Console.WriteLine($"   1. Deploy agent to Container Apps");
            Console.WriteLine($"   2. Configure managed identity permissions");
            Console.WriteLine($"   3. Register connector in Logic Apps");
            Console.WriteLine($"   4. Submit to Azure Marketplace (optional)");

            return endpointConfig;
        }

        /// <summary>
        /// Generates a deployment script for Container Apps.
        /// </summary>
        /// <param name="agentName">Name of the agent</param>
        /// <param name="endpointConfig">Endpoint configuration</param>
        /// <returns>Deployment script content</returns>
        public string GenerateDeploymentScript(string agentName, Dictionary<string, object> endpointConfig)
        {
            var endpoint = (Dictionary<string, object>)endpointConfig["endpoint"];
            var identity = (Dictionary<string, object>)endpointConfig["identity"];

            var script = $@"#!/bin/bash
# Deployment script for {agentName}

# Variables
RESOURCE_GROUP=""your-resource-group""
LOCATION=""eastus""
CONTAINER_APP_NAME=""{agentName.ToLower().Replace(" ", "-")}""
CONTAINER_APP_ENV=""agent-env""
IMAGE=""mcr.microsoft.com/azure-ai/agents:latest""

# Create Container Apps environment
echo ""Creating Container Apps environment...""
az containerapp env create \
  --name $CONTAINER_APP_ENV \
  --resource-group $RESOURCE_GROUP \
  --location $LOCATION

# Create Container App with managed identity
echo ""Creating Container App...""
az containerapp create \
  --name $CONTAINER_APP_NAME \
  --resource-group $RESOURCE_GROUP \
  --environment $CONTAINER_APP_ENV \
  --image $IMAGE \
  --target-port 8080 \
  --ingress external \
  --system-assigned

# Get managed identity principal ID
PRINCIPAL_ID=$(az containerapp show \
  --name $CONTAINER_APP_NAME \
  --resource-group $RESOURCE_GROUP \
  --query identity.principalId -o tsv)

echo ""Managed Identity Principal ID: $PRINCIPAL_ID""

# Assign permissions (example for AI Foundry)
echo ""Assigning permissions...""
az role assignment create \
  --assignee $PRINCIPAL_ID \
  --role ""Cognitive Services User"" \
  --scope /subscriptions/$SUBSCRIPTION_ID/resourceGroups/$RESOURCE_GROUP

echo ""Deployment complete!""
echo ""Endpoint URL: {endpoint["url"]}""
";

            return script;
        }

        /// <summary>
        /// Validates endpoint configuration for production readiness.
        /// </summary>
        /// <param name="endpointConfig">Endpoint configuration to validate</param>
        /// <returns>Validation results</returns>
        public Dictionary<string, object> ValidateEndpointConfiguration(Dictionary<string, object> endpointConfig)
        {
            var issues = new List<string>();
            var warnings = new List<string>();

            var endpoint = (Dictionary<string, object>)endpointConfig["endpoint"];
            var identity = (Dictionary<string, object>)endpointConfig["identity"];
            var apiManagement = (Dictionary<string, object>)endpointConfig["api_management"];

            // Check authentication
            if (endpoint["authentication"].ToString() != "managed_identity")
            {
                warnings.Add("Managed identity authentication is recommended for production");
            }

            // Check API management
            if (!(bool)apiManagement["enabled"])
            {
                warnings.Add("API Management should be enabled for production");
            }

            // Check rate limiting
            var rateLimiting = (Dictionary<string, object>)apiManagement["rate_limiting"];
            if ((int)(long)rateLimiting["requests_per_minute"] < 10)
            {
                issues.Add("Rate limit too restrictive for production use");
            }

            // Check CORS
            var cors = (Dictionary<string, object>)apiManagement["cors"];
            if (!(bool)cors["enabled"])
            {
                warnings.Add("CORS should be configured for web applications");
            }

            var isValid = issues.Count == 0;

            return new Dictionary<string, object>
            {
                ["is_valid"] = isValid,
                ["issues"] = issues,
                ["warnings"] = warnings,
                ["timestamp"] = DateTime.UtcNow.ToString("O")
            };
        }

        /// <summary>
        /// Generates API documentation for the production endpoint.
        /// </summary>
        /// <param name="agentName">Name of the agent</param>
        /// <param name="endpointConfig">Endpoint configuration</param>
        /// <returns>API documentation in markdown format</returns>
        public string GenerateAPIDocumentation(string agentName, Dictionary<string, object> endpointConfig)
        {
            var endpoint = (Dictionary<string, object>)endpointConfig["endpoint"];
            var logicApps = (Dictionary<string, object>)endpointConfig["logic_apps"];

            var doc = $@"# {agentName} API Documentation

## Overview
Production endpoint for {agentName} agent.

## Base URL
```
{endpoint["url"]}
```

## Authentication
Type: {endpoint["authentication"]}

Required Headers:
- `Authorization`: Bearer token or managed identity
- `Content-Type`: application/json

## Endpoints

### POST /api/query
Query the agent with a question.

**Request Body:**
```json
{{
  ""question"": ""Your question here"",
  ""conversation_id"": ""optional-conversation-id"",
  ""user_id"": ""optional-user-id""
}}
```

**Response:**
```json
{{
  ""response"": ""Agent's response"",
  ""conversation_id"": ""conversation-id"",
  ""status"": ""completed"",
  ""metadata"": {{
    ""latency_ms"": 1234,
    ""tokens_used"": 567
  }}
}}
```

### GET /health
Health check endpoint.

**Response:**
```json
{{
  ""status"": ""healthy"",
  ""timestamp"": ""2025-10-23T19:00:00Z""
}}
```

## Logic Apps Integration

### Available Actions
";

            var actions = (string[])logicApps["actions"];
            foreach (var action in actions)
            {
                doc += $"- `{action}`\n";
            }

            doc += "\n### Available Triggers\n";
            var triggers = (string[])logicApps["triggers"];
            foreach (var trigger in triggers)
            {
                doc += $"- `{trigger}`\n";
            }

            doc += @"

## Rate Limits
- 100 requests per minute
- 5000 requests per hour

## Error Codes
- `400`: Bad Request - Invalid input
- `401`: Unauthorized - Missing or invalid authentication
- `429`: Too Many Requests - Rate limit exceeded
- `500`: Internal Server Error - Agent processing error

## Support
For support and questions, contact: support@example.com
";

            return doc;
        }
    }
}

File: samples/microsoft/java/enterprise-agent-tutorial/2-prototype-to-production/src/main/java/com/microsoft/azure/samples/ProductionDeployment.java
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

package com.microsoft.azure.samples;

import com.azure.ai.agents.models.AgentVersionObject;
import com.google.gson.Gson;

import java.io.FileWriter;
import java.io.IOException;
import java.time.Instant;
import java.util.*;

/**
 * Handles production endpoint configuration and deployment.
 * Generates configuration for Azure Container Apps, Marketplace, and Logic Apps.
 */
public class ProductionDeployment {
    
    private final Gson gson;
    
    public ProductionDeployment(Gson gson) {
        this.gson = gson;
    }
    
    /**
     * Generate production endpoint configuration for the agent.
     * 
     * @param agent The agent to configure for production
     */
    public void generateProductionEndpointConfig(AgentVersionObject agent) {
        System.out.println("\n🚀 PRODUCTION ENDPOINT DEPLOYMENT");
        System.out.println("=".repeat(70));

        // <endpoint_configuration>
        // NOTE: This code is a non-runnable snippet of the larger sample code from which it is taken.
        Map<String, Object> endpointConfig = new HashMap<>();

        // Container App Configuration
        Map<String, Object> containerApp = new HashMap<>();
        containerApp.put("name", "workplace-assistant");
        containerApp.put("resource_group", "agents-production-rg");
        containerApp.put("environment", "agents-prod-env");
        containerApp.put("container_image", "your-registry.azurecr.io/workplace-assistant:latest");
        containerApp.put("ingress_type", "external");
        containerApp.put("target_port", 8000);
        containerApp.put("min_replicas", 2);
        containerApp.put("max_replicas", 10);

        // Marketplace Configuration
        Map<String, Object> marketplace = new HashMap<>();
        marketplace.put("offer_id", "workplace-assistant");
        marketplace.put("plan_id", "enterprise");
        marketplace.put("publisher_id", "contoso");
        marketplace.put("categories", Arrays.asList("AI + Machine Learning", "Productivity"));
        marketplace.put("pricing_model", "per_user_per_month");

        // Logic Apps Connector Configuration
        Map<String, Object> logicAppsConnector = new HashMap<>();
        logicAppsConnector.put("connector_name", "WorkplaceAssistant");
        logicAppsConnector.put("display_name", "Workplace Assistant Connector");
        logicAppsConnector.put("description", "Connect to your workplace assistant for automated workflows");
        logicAppsConnector.put("icon_url", "https://your-cdn.com/icon.png");

        // Managed Identity Configuration
        Map<String, Object> managedIdentity = new HashMap<>();
        managedIdentity.put("type", "SystemAssigned");
        managedIdentity.put("role_assignments", Arrays.asList(
                createRoleAssignment("Cognitive Services User", "AI Foundry Project"),
                createRoleAssignment("Storage Blob Data Reader", "Storage Account")
        ));

        // API Management Configuration
        Map<String, Object> apim = new HashMap<>();
        apim.put("api_name", "workplace-assistant-api");
        apim.put("api_path", "workplace-assistant");
        apim.put("rate_limit", 100);
        apim.put("quota_limit", 10000);
        apim.put("cors_enabled", true);
        apim.put("cors_origins", Arrays.asList("https://your-app.com"));
        // </endpoint_configuration>

        endpointConfig.put("configuration_date", Instant.now().toString());
        endpointConfig.put("agent_id", agent.getId());
        endpointConfig.put("agent_name", agent.getName());
        endpointConfig.put("container_app", containerApp);
        endpointConfig.put("marketplace", marketplace);
        endpointConfig.put("logic_apps_connector", logicAppsConnector);
        endpointConfig.put("managed_identity", managedIdentity);
        endpointConfig.put("api_management", apim);

        saveToFile(endpointConfig, "endpoint_config.json");

        // Generate Logic Apps definition
        Map<String, Object> logicAppDef = new HashMap<>();
        logicAppDef.put("$schema", "https://schema.management.azure.com/providers/Microsoft.Logic/schemas/2016-06-01/workflowdefinition.json#");
        logicAppDef.put("contentVersion", "1.0.0.0");
        
        Map<String, Object> parameters = new HashMap<>();
        Map<String, Object> triggers = new HashMap<>();
        triggers.put("manual", createManualTrigger());
        
        Map<String, Object> actions = new HashMap<>();
        actions.put("Call_Workplace_Assistant", createHttpAction(
                "POST",
                "https://workplace-assistant.azurecontainerapps.io/api/chat",
                "@triggerBody()"
        ));
        
        Map<String, Object> definition = new HashMap<>();
        definition.put("triggers", triggers);
        definition.put("actions", actions);
        
        logicAppDef.put("parameters", parameters);
        logicAppDef.put("definition", definition);

        saveToFile(logicAppDef, "logic_app_definition.json");

        System.out.println("✅ Production endpoint configured successfully");
        System.out.println("   Deployment: Azure Container Apps");
        System.out.println("   Authentication: Managed Identity");
        System.out.println("   API Management: Rate limiting enabled");
        System.out.println("   Configuration saved: endpoint_config.json");
        System.out.println("   Logic Apps definition: logic_app_definition.json");
    }
    
    private Map<String, Object> createRoleAssignment(String role, String scope) {
        Map<String, Object> assignment = new HashMap<>();
        assignment.put("role", role);
        assignment.put("scope", scope);
        return assignment;
    }
    
    private Map<String, Object> createManualTrigger() {
        Map<String, Object> trigger = new HashMap<>();
        trigger.put("type", "Request");
        trigger.put("kind", "Http");
        Map<String, Object> inputs = new HashMap<>();
        inputs.put("method", "POST");
        trigger.put("inputs", inputs);
        return trigger;
    }
    
    private Map<String, Object> createHttpAction(String method, String uri, String body) {
        Map<String, Object> action = new HashMap<>();
        action.put("type", "Http");
        Map<String, Object> inputs = new HashMap<>();
        inputs.put("method", method);
        inputs.put("uri", uri);
        inputs.put("body", body);
        action.put("inputs", inputs);
        return action;
    }
    
    private void saveToFile(Object data, String filename) {
        try {
            FileWriter writer = new FileWriter(filename);
            gson.toJson(data, writer);
            writer.close();
        } catch (IOException e) {
            System.err.println("Error saving file " + filename + ": " + e.getMessage());
        }
    }
}
